polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp

```diff
@@ -703,13 +710,218 @@ void AffineParallelGuardBuilder::getConstraints(
     }
 }
 
+//===----------------------------------------------------------------------===//
+// SYCLAccessorVersionBuilder
+//===----------------------------------------------------------------------===//
+
+// TODO: Support other kind of loops.
+class SYCLAccessorVersionBuilder : public SCFLoopVersionBuilder {
+public:
+  SYCLAccessorVersionBuilder(LoopLikeOpInterface loop)
+      : SCFLoopVersionBuilder(loop) {}
+
+  bool isVersionCandidate(const Operation &op1, const Operation &op2) {
+    if (!EnableSYCLAccessorVersioning)
+      return false;
+
+    // Currently only allow one accessor pair.
+    if (!empty())
+      return false;
+
+    auto getMemrefOp = [](const Operation &op) -> Operation * {
+      if (auto loadOp = dyn_cast<AffineLoadOp>(op))
+        return loadOp.getMemref().getDefiningOp();
+      if (auto storeOp = dyn_cast<AffineStoreOp>(op))
+        return storeOp.getMemref().getDefiningOp();
+      return nullptr;
+    };
+
+    auto accSub1 =
+        dyn_cast_or_null<sycl::SYCLAccessorSubscriptOp>(getMemrefOp(op1));
+    auto accSub2 =
+        dyn_cast_or_null<sycl::SYCLAccessorSubscriptOp>(getMemrefOp(op2));
+    if (!accSub1 || !accSub2)
+      return false;
+
+    TypedValue<MemRefType> acc1 = accSub1.getAcc();
+    TypedValue<MemRefType> acc2 = accSub2.getAcc();
+    if (acc1 == acc2)
+      return false;
+
+    if (!loop.isDefinedOutsideOfLoop(acc1) ||
+        !loop.isDefinedOutsideOfLoop(acc2))
+      return false;
+
+    AccessorPair = std::pair(acc1, acc2);
+
+    return true;
+  }
+
+  void versionLoop() {
+    if (empty())
+      return;
```

Comment:
Maybe assert?

Reply:
We should just not version if there is not candidates. IMO, shouldn't assert.